eboutic/forms.py

```diff
@@ -0,0 +1,168 @@
+    the user submits his basket from /eboutic/
+
+    Because it must check an unknown number of fields, coming from a cookie
+    and needing some databases checks to be performed, inheriting from forms.Form
+    or using formset would have been likely to end in a big ball of wibbly-wobbly hacky stuff.
+    Thus this class is a pure standalone and performs its operations by its own means.
+    However, it still tries to share some similarities with a standard django Form.
+
+    Example:
+        ::
+
+            def my_view(request):
+                form = BasketForm(request)
+                form.clean()
+                if form.is_valid():
+                    # perform operations
+                else:
+                    errors = form.get_error_messages()
+
+                    # return the cookie that was in the request, but with all
+                    # incorrects elements removed
+                    cookie = form.get_cleaned_cookie()
+
+    You can also use a little shortcut by directly calling `form.is_valid()`
+    without calling `form.clean()`. In this case, the latter method shall be
+    implicitly called.
+    """
+
+    # check the json is an array containing non-nested objects.
+    # values must be strings or numbers
+    # this is matched :
+    # [{"id": 4, "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    # and this is not :
+    # [{"id": {"nested_id": 10}, "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    # and neither does this :
+    # [{"id": ["nested_id": 10], "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    json_cookie_re = re.compile(
+        r"^\[\s*(\{\s*(\"[^\"]*\":\s*(\"[^\"]{0,64}\"|\d{0,5}\.?\d+),?\s*)*\},?\s*)*\s*\]$"
+    )
+
+    def __init__(self, request: HttpRequest):
+        self.user = request.user
+        self.cookies = request.COOKIES
+        self.error_messages = set()
+        self.correct_cookie = []
+
+    def clean(self) -> None:
+        """
+        Perform all the check, but return nothing.
+        To know if the form is valid, the `is_valid()` method must be used.
+
+        The form shall be considered as valid if it meets all the following conditions :
+            - it contains a "basket_items" key in the cookies of the request given in the constructor
+            - this cookie is a list of objects formatted this way : `[{'id': <int>, 'quantity': <int>,
+             'name': <str>, 'unit_price': <float>}, ...]`. The order of the fields in each object does not matter
+            - all the ids are positive integers
+            - all the ids refer to products available in the EBOUTIC
+            - all the ids refer to products the user is allowed to buy
+            - all the quantities are positive integers
+        """
+        basket = self.cookies.get("basket_items", None)
+        if basket is None or basket in ("[]", ""):
+            self.error_messages.add("You have no basket")
```

Comment:
Pour ce genre de message d'erreur, tu devrais utiliser gettext ou gettext_lazy pour utiliser les traductions

Reply:
`from django.utils.translation import gettext_lazy as _`
ou `from django.utils.translation import gettext as _`
Et après tu peux faire `self.error_messages.add(_("You have no basket"))`
Comme ça t'as déjà un truc déjà traduit sur le frontend, parce que tu peux pas appeler les traductions quand t'es pas dans la génération du template alors c'est maintenant qu'il faut le faire

Reply:
https://github.com/imperosol/sith3/blob/e20375cdbc08c260dd210f0a72d2f66eadb81e3e/eboutic/forms.py#L30
https://github.com/imperosol/sith3/blob/e20375cdbc08c260dd210f0a72d2f66eadb81e3e/eboutic/forms.py#L162